File: tools/packager/logging.cc
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "sky/tools/packager/logging.h"

#include "base/logging.h"

bool LogIfError(Dart_Handle handle) {
  if (Dart_IsError(handle)) {
    LOG(ERROR) << Dart_GetError(handle);

    // Only unhandled exceptions have stacktraces.
    if (!Dart_ErrorHasException(handle))
      return true;

    Dart_Handle stacktrace = Dart_ErrorGetStacktrace(handle);
    const char* stacktrace_cstr = "";
    Dart_StringToCString(Dart_ToString(stacktrace), &stacktrace_cstr);
    LOG(ERROR) << stacktrace_cstr;
    return true;
  }
  return false;
}

std::string StringFromDart(Dart_Handle string) {
  CHECK(Dart_IsString(string));
  uint8_t* utf8_array;
  intptr_t length;
  Dart_StringToUTF8(string, &utf8_array, &length);
  return std::string(reinterpret_cast<const char*>(utf8_array), length);
}

Dart_Handle StringToDart(const std::string& string) {
  return Dart_NewStringFromUTF8(reinterpret_cast<const uint8_t*>(string.data()),
                                string.length());
}

File: .pub-cache/hosted/pub.dartlang.org/dds-2.2.5/lib/src/dap/logging.dart
// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'package:vm_service/vm_service.dart' as vm;

typedef Logger = void Function(String);

/// Wraps a [Logger] as a [vm/Log] to be passed to the VM Service library.
class VmServiceLogger extends vm.Log {
  final Logger _logger;

  VmServiceLogger(this._logger);

  @override
  void severe(String message) => _logger.call('ERROR: $message');

  @override
  void warning(String message) => _logger.call('WARN: $message');
}
